pkg/client/client.go

```diff
@@ -1244,3 +1322,80 @@ func check(dep pkg.Dependency, newDepPath string) bool {
 
 	return dep.Sum == sum
 }
+
+func FindSource[K comparable, T any](g graph.Graph[K, T]) ([]K, error) {
+	if !g.Traits().IsDirected {
+		return nil, fmt.Errorf("cannot find source of a non-DAG graph ")
+	}
+
+	predecessorMap, err := g.PredecessorMap()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get predecessor map: %w", err)
+	}
+
+	var sources []K
+	for vertex, predecessors := range predecessorMap {
+		if len(predecessors) == 0 {
+			sources = append(sources, vertex)
+		}
+	}
+	return sources, nil
+}
+
+// Union combines two given graphs into a new graph. The vertex hashes in both
+// graphs are expected to be unique. The two input graphs will remain unchanged.
+//
+// Both graphs should be either directed or undirected. All traits for the new
+// graph will be derived from g.
+// 
+// If the same vertex/edge happens to be in both g and h, then an error will not be 
+// thrown as happens in original Union function and successful operation takes place.
+func Union[K comparable, T any](g, h graph.Graph[K, T]) (graph.Graph[K, T], error) {
```

Comment:
Move graph related functions to graph.go or dep_graph.go

Reply:
Done! 
I have also added unit test